### PR TITLE
change staking provider owner

### PR DESCRIPTION
**Pull request by sasurobert**
change staking provider owner

### CODE REVIEW

File: vm/systemSmartContracts/delegation.go [hunk 1]
```diff
@@ -929,6 +935,30 @@ func (d *delegation) checkBLSKeysIfExistsInStakingSC(blsKeys [][]byte) bool {
...
+	if len(args.Arguments) != 1 {
+		d.eei.AddReturnMessage("wrong number of arguments, expected 1")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments[0]) != len(args.CallerAddr) {
+		d.eei.AddReturnMessage("invalid argument, wanted an address")
+		return vmcommon.UserError
+	}
+
+	d.eei.SetStorage([]byte(ownerKey), args.Arguments[0])
```
Comment (raduchis) [inline vm/systemSmartContracts/delegation.go:957]: what happens to the initialOwnerFunds, 1250 EGLD, of the previous owner? When can he withdraw?

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

File: vm/systemSmartContracts/delegation.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:957]: he cannot withdraw - that is for the new owner.

File: vm/systemSmartContracts/delegation.go [hunk 2]
```diff
@@ -929,6 +935,30 @@ func (d *delegation) checkBLSKeysIfExistsInStakingSC(blsKeys [][]byte) bool {
 	return false
 }
 
+func (d *delegation) changeOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:938]: 1. Actually we do not let this through delegation manager, so will change.
2. No, only one address.
3. Yes, you can change to a SC address. It is the SC developers job to implement stuff.

[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:938]: That is another usecase, I do not see why should we complicate this.

File: vm/systemSmartContracts/delegation.go [hunk 3]
```diff
@@ -954,6 +954,13 @@ func (d *delegation) changeOwner(args *vmcommon.ContractCallInput) vmcommon.Retu
 		return vmcommon.UserError
 	}
 
+	dataFromDelegationMGR := d.eei.GetStorageFromAddress(d.delegationMgrSCAddress, args.Arguments[0])
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:957]: please rename this to: `dataFromDelegationManager`

File: vm/systemSmartContracts/delegation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/delegation.go:938]: ok

File: vm/systemSmartContracts/delegation.go (hunk 1 above)
Comment (raduchis) [reply vm/systemSmartContracts/delegation.go:964]: ok... so during the offchain agreement to change the owner, the new owner should reimburse the old owner the initialOwnerFunds?

File: vm/systemSmartContracts/delegation_test.go [hunk 4]
```diff
@@ -5322,3 +5322,74 @@ func TestDelegation_correctNodesStatus(t *testing.T) {
...
+
+	d, _ := NewDelegationSystemSC(args)
+	vmInput := getDefaultVmInputForFunc("changeOwner", vmInputArgs)
+	vmInput.CallValue = big.NewInt(0)
+	vmInput.CallerAddr = delegationsMap[ownerKey]
+	vmInput.Arguments = append(vmInput.Arguments, []byte("second123"))
+
+	returnCode := d.Execute(vmInput)
+	assert.Equal(t, returnCode, vmcommon.Ok)
+	assert.Equal(t, delegationsMap[ownerKey], []byte("second123"))
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply vm/systemSmartContracts/delegation_test.go:5403]: still not done

[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation_test.go:5403]: done

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by iulianpascalau — DISMISSED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

File: vm/systemSmartContracts/delegation.go [hunk 5]
```diff
@@ -960,6 +960,33 @@ func (d *delegation) changeOwner(args *vmcommon.ContractCallInput) vmcommon.Retu
 		return vmcommon.UserError
 	}
 
+	isNew, _, err := d.getOrCreateDelegatorData(args.Arguments[0])
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !isNew {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:968]: :+1:

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Pull request merged**